pkgs/development/interpreters/cling/flags.nix

```diff
@@ -0,0 +1,28 @@
+{cling, llvmPackages_5, glibc}:
+
+# This file contains a list of flags used to "isolate" a Cling installation.
+# This means
+# a) prevent it from searching for system include files and libs
+# b) provide it with the include files and libs it needs (C and C++ standard library)
+
+# This is provided as a separate file because it's handy to be able to be able to pass them
+# to tools that wrap Cling, particularly Jupyter kernels such as xeus-cling and the built-in
+# jupyter-cling-kernel.
+
+# Both of these use Cling as a library by linking against libclingJupyter.so, so the
+# makeWrapper approach to wrapping the binary doesn't work.
+# Thus, if you're packaging a Jupyter kernel, you either need to pass these flags as extra
+# args to xcpp (for xeus-cling) or put them in the environment variable CLING_OPTS
+# (for jupyter-cling-kernel)
+
+[
+  "-nostdinc"
+  "-nostdinc++"
+  "-I" "${cling}/include"
+  "-I" "${glibc.dev}/include"
+  "-I" "${llvmPackages_5.libcxx}/include/c++/v1"
+  "-I" "${cling}/lib/clang/5.0.2/include"
+  "-L" "${cling}/lib"
+  "-L" "${llvmPackages_5.libcxx}/lib"
+  "-l" "${llvmPackages_5.libcxx}/lib/libc++.so"
```

Comment:
One question here for @Axel-Naumann, if you wouldn't mind. Do you happen to know why this line seems to be necessary? I'm trying a simple example where I type `#include <iostream>` and then `std::cout << "HI";`. If I remove the `-l` line here, I get an error about the shared library not being present. I would have expected Cling to auto-load the C++ standard library, am I missing something? Thanks!

Reply:
Absolutely! Sp NixOS uses libc++? What is the error message you see without `-l libc++.so`?

Does it mention "missing symbol XYZ"? What does this show:
```
gdb cling
cling$ dlsym(0, "XYZ")
^Z
(gdb) info sharedlibrary
```
Thanks!

Reply:
> Sp NixOS uses libc++?

Nix can provide pretty much any version you like of the standard libraries. I'm passing `-nostdinc` and `-nostdinc++` so that I can precisely control what gets passed in. In the code I linked, I pass `-I` flags with GNU libc and LLVM 5 libc++, plus a few others.

> What is the error message you see without `-l libc++.so`? Does it mention "missing symbol XYZ"?

I get a bunch of them saying "symbol unresolved while linking," see here:

<details><summary>Error messages</summary>
<p>

```
IncrementalExecutor::executeFunction: symbol '_ZNSt3__15ctypeIcE2idE' unresolved while linking [cling interface function]!
You are probably missing the definition of std::__1::ctype<char>::id
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZNSt3__113basic_ostreamIcNS_11char_traitsIcEEE6sentryC1ERS3_' unresolved while linking [cling interface function]!
You are probably missing the definition of std::__1::basic_ostream<char, std::__1::char_traits<char> >::sentry::sentry(std::__1::basic_ostream<char, std::__1::char_traits<char> >&)
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE6__initEmc' unresolved while linking [cling interface function]!
You are probably missing the definition of std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__init(unsigned long, char)
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZNSt3__113basic_ostreamIcNS_11char_traitsIcEEE6sentryD1Ev' unresolved while linking [cling interface function]!
You are probably missing the definition of std::__1::basic_ostream<char, std::__1::char_traits<char> >::sentry::~sentry()
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZNKSt3__16locale9use_facetERNS0_2idE' unresolved while linking [cling interface function]!
You are probably missing the definition of std::__1::locale::use_facet(std::__1::locale::id&) const
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZNSt3__16localeD1Ev' unresolved while linking [cling interface function]!
You are probably missing the definition of std::__1::locale::~locale()
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEED1Ev' unresolved while linking [cling interface function]!
You are probably missing the definition of std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::~basic_string()
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZNSt3__14coutE' unresolved while linking [cling interface function]!
You are probably missing the definition of std::__1::cout
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZNKSt3__18ios_base6getlocEv' unresolved while linking [cling interface function]!
You are probably missing the definition of std::__1::ios_base::getloc() const
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZNSt3__18ios_base5clearEj' unresolved while linking [cling interface function]!
You are probably missing the definition of std::__1::ios_base::clear(unsigned int)
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZNSt3__18ios_base33__set_badbit_and_consider_rethrowEv' unresolved while linking [cling interface function]!
You are probably missing the definition of std::__1::ios_base::__set_badbit_and_consider_rethrow()
Maybe you need to load the corresponding shared library?
```

</p>
</details>

> What does this show:

Picking a random symbol from the errors: 

```
[cling]$ #include <iostream>
[cling]$ #include <dlfcn.h>
[cling]$ dlsym(0, "_ZNSt3__15ctypeIcE2idE")
(void *) nullptr
```

After attaching GDB to the process, I see libc but not libc++ present. But maybe my `dlsym` call didn't work.

<details><summary>GDB output</summary>
<p>

```
(gdb) info sharedlibrary
From                To                  Syms Read   Shared Object Library
0x00007fb945eb75a0  0x00007fb945ec5dd1  Yes (*)     /nix_frozen/store/kng3jbgmpdl8g26jjw2bwy19djffna5s-glibc-2.30/lib/libpthread.so.0
0x00007fb945ea93a0  0x00007fb945eac14c  Yes (*)     /nix_frozen/store/kng3jbgmpdl8g26jjw2bwy19djffna5s-glibc-2.30/lib/librt.so.1
0x00007fb945ea3120  0x00007fb945ea3dd5  Yes (*)     /nix_frozen/store/kng3jbgmpdl8g26jjw2bwy19djffna5s-glibc-2.30/lib/libdl.so.2
0x00007fb945e49ce0  0x00007fb945e849f8  Yes (*)     /nix_frozen/store/42d5bjsnjwkwwr8b9ri1m0wkaw6gdhw5-ncurses-6.1-20190112/lib/libncursesw.so.6
0x00007fb945e17310  0x00007fb945e27cfa  Yes (*)     /nix_frozen/store/rfrn71d1dy43cw1gg6pm7xj72p37dd5z-zlib-1.2.11/lib/libz.so.1
0x00007fb945cde1d0  0x00007fb945d74e12  Yes (*)     /nix_frozen/store/kng3jbgmpdl8g26jjw2bwy19djffna5s-glibc-2.30/lib/libm.so.6
0x00007fb945cbb2e0  0x00007fb945ccbb9d  Yes (*)     /nix_frozen/store/kng3jbgmpdl8g26jjw2bwy19djffna5s-glibc-2.30/lib/libgcc_s.so.1
0x00007fb945b1b340  0x00007fb945c5e18b  Yes (*)     /nix_frozen/store/kng3jbgmpdl8g26jjw2bwy19djffna5s-glibc-2.30/lib/libc.so.6
0x00007fb945ed5090  0x00007fb945ef2b90  Yes (*)     /nix_frozen/store/kng3jbgmpdl8g26jjw2bwy19djffna5s-glibc-2.30/lib/ld-linux-x86-64.so.2
```

</p>
</details>

Reply:
cling expects that it is build ABI compatible with what it interprets. If cling is built with libc++, #including its headers is fine. If cling is built with libstdc++, #including headers from libc++ is... - not tested, to say the least :-) Is cling linked against libstdc++? `dlsym` seems to suggest so.

Reply:
Hmm. I'm using Nix's `clangStdenv` to build everything, so I'm assuming it's built against `libc++`. I'll double-check that. What makes you think it's linked against `libstdc++`? Can you tell that from the name mangling?

When I run `ldd cling`, I actually don't see any `*c++` library linked against it. I only see `libncursesw` and several `glibc` dependencies. So I guess the C++ lib is statically linked?

Reply:
Aha, I think I confirmed it. When I pass the C++ headers that ship with GCC instead, then it works without any `-l` line. Thanks!

I guess the next question is, *should* I link against `libc++` instead? It seems weird to have everything be LLVM except the C++ lib. But when I tried to switch to `libcxxStdenv` I got linker errors when building Cling. Is one or the other preferred?

Reply:
Well, there's lldb coming with llvm, but that's not necessarily a reason to use it with cling :-) If you build cling with libstdc++ (GCC's standard lib) then you should also use its headers. Problem solved?

Reply:
> What makes you think it's linked against libstdc++? Can you tell that from the name mangling?

Yes, else the libc++ symbols would be found in cling. OTOH if you #include gcc libstdc++ headers they will be found; try
```
#include <dlfcn.h>
dlsym(0, "_ZNSi3getEv")
```
If libstdc++.so is statically linked then some C++ stdlib symbols might be missing.

And you should link against whatever is the system default stdlib.

Reply:
AFAIK Nix doesn't have a default stdlib--you can build different packages against different ones if you wish.

I think we want this packaged Cling to be as general-purpose as possible. If users want to work on C++ targeted at GCC vs. LLVM then the chosen stdlib could matter, right? I'm wondering if it would make sense to make this configurable, so users could choose either the "GCC flavor" or "LLVM flavor" to build. What do you think @veprbl ?

Reply:
I think we should just build cling with the `stdenv` so that it stays compatible with the rest of the package set. I don't yet see how `llvmPackages_5.libcxx` fits into the picture.